### PR TITLE
[mlir][Affine] Split off delinearize parts that depend on last component

### PR DESCRIPTION
**Pull request by krzysz00**
If we have

    %0 = affine.linearize_index disjoint [%a, %b] by (A, B)
    %1:3 = affine.delinearize_index %0 into (A, B1, B2)

where B = B1 * B2 (or some mor complex product), we can simplify this to

    %0 = affine.linearize_index disjoint [%a] by (A)
    %1a:1 = affine.delinearize_index %0 into (A)
    %1b:2 = affine.delinearize_index %b into (B1, B2)

This, and more complex cases, prevent us from adding terms together only to divide them away from each other.

### CODE REVIEW

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp [hunk 1]
```diff
@@ -4729,12 +4729,97 @@ struct CancelDelinearizeOfLinearizeDisjointExactTail
...
+    }
+
+    if (sizeToSplit < target)
+      return rewriter.notifyMatchFailure(
+          delinearizeOp, "product of known basis elements doesn't exceed last "
+                         "linearize argument");
+
+    if (elemsToSplit < 2)
+      return rewriter.notifyMatchFailure(
+          delinearizeOp, "don't have a non-trivial basis product");
```
Comment (Abhishek-Varma) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4791]:
```suggestion
          delinearizeOp, "need at least two elements to form the basis product");
```

Comment (Abhishek-Varma) [inline mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4791]: Also, maybe add a lit test to test this path as well.

**Review by Abhishek-Varma — APPROVED** [review]
Amazing PR @krzysz00 ! Approving with two requests! Thanks!

File: mlir/lib/Dialect/Affine/IR/AffineOps.cpp (hunk 1 above)
Comment (krzysz00) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp:4791]: I think it's implicit in an existing test that permutes but I'll go add another one if it isn't.

Comment (krzysz00) [reply mlir/lib/Dialect/Affine/IR/AffineOps.cpp]: ... Ok, yeah, the case where the final components agree exactly gets sent down a different pattern, is what's going on

**Pull request merged**